syntax = "proto3";

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);
  rpc get(GetRequest) returns (GetResponse);
  rpc slowDown(SlowDownRequest) returns (SlowDownResponse);
  rpc restore(Empty) returns (RestoreResponse);

  // 2PC protocol methods
  rpc canCommit(CanCommitRequest) returns (CanCommitResponse);
  rpc doCommit(DoCommitRequest) returns (DoCommitResponse);
  rpc abort(AbortRequest) returns (AbortResponse);
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
}

message SlowDownRequest {
  int32 seconds = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message RestoreRequest {

}

message RestoreResponse {
  bool success = 1;
}

message Empty {

}

message CanCommitRequest {
  string key = 1;
  string value = 2;
}

message CanCommitResponse {
  bool canCommit = 1;
}

message DoCommitRequest {
  string key = 1;
  string value = 2;
}

message DoCommitResponse {
  bool success = 1;
}

message AbortRequest {
  string key = 1;
}

message AbortResponse {
  bool success = 1;
}
